Add search-router extension

franzwilhelm opened the pull request:
## Description

Search Router implements DuckDuckGo-style bangs search directly in Raycast without sending traffic through third-party websites. It allows users to search specific websites using shortcuts like:

- `!g cats` to search Google for "cats"
- `!w cats` to search Wikipedia for "cats" 
- `Help me fix my code !t3` to use t3.chat for AI code help

For the best experience, users can set up Search Router as a Fallback Command in Raycast, enabling bang searches directly from the main Raycast search.

Made this[ based on inspiration from Theo Browne on YouTube](https://www.youtube.com/watch?v=_DnNzRaBWUU), and I want his take on the PR if he has the time to review it.

## Screencast


### Using the search command
https://github.com/user-attachments/assets/9cf680ea-a841-4450-aaba-371fba3b84c5

### Browsing available search engines
https://github.com/user-attachments/assets/9cb6341f-2244-44e1-bd3e-9abe414e6518


## Checklist

- [x] I read the [extension guidelines](https://developers.raycast.com/basics/prepare-an-extension-for-store)
- [x] I read the [documentation about publishing](https://developers.raycast.com/basics/publish-an-extension)
- [x] I ran `npm run build` and [tested this distribution build in Raycast](https://developers.raycast.com/basics/prepare-an-extension-for-store#metadata-and-configuration)
- [x] I checked that files in the `assets` folder are used by the extension itself
- [x] I checked that assets used by the `README` are placed outside of the `metadata` folder

pernielsentikaer reviewed (changes requested):
Hi 👋 

Thanks for your contribution 💪 

I have now tested your extension, and I have some feedback ready for you:

- I'm not sure I understand why we need to use the preferences API here, why not use a simple `useCachedState`

- Should we consider adding a `Pin` function with the ability to move them up and down too?

I'm looking forward to testing this extension again 🔥 

Request a new review when you are ready. Feel free to contact me here or at [Slack](https://www.raycast.com/community) if you have any questions.

pernielsentikaer commented:
@greptileai please check again

pernielsentikaer commented on extensions/search-router/src/search.tsx:
```diff
@@ -0,0 +1,33 @@
+
+    const cleanQuery = query.replace(/!\S+\s*/i, "").trim();
+    const searchUrl = searchEngine.u.replace("{{{s}}}", encodeURIComponent(cleanQuery).replace(/%2F/g, "/"));
+    await open(searchUrl);
```
That's not a good solution 😊

franzwilhelm commented:
@pernielsentikaer moved over to `useCachedState`. Note that I had to use the Cache API directly in addition to the hook, as the main search command is in mode `no-view`, meaning react hooks are unavailable.

lacherogwu commented:
Looks good! What's the final name: Search Router or Search Plus? If it's Search Router, we need to update the photos. I've also noticed that the SQLite database file still exists in the assets; you can remove it.

pernielsentikaer commented:
There is something with our publish command that doesn't catch deletions correctly, so I deleted it. If you're having issues, @franzwilhelm, from now on, just ping me on Slack, and I can make the corrections.

franzwilhelm commented:
Appreciated @pernielsentikaer! I managed to get to a state where I can git push directly without ray publish as I too noticed the bug with deletions.

@lacherogwu Again thank you for noticing stuff like this! Updated all asset names and screenshots now. Also added the unduck keyword to both commands, so users can find it more easily if they forget the name search router, but remember the name unduck.

pernielsentikaer reviewed (approved):
It search keywords from package.json, let's get the first version live 😊
